File: sources/scalac/util/AbstractFileReader.java
/*     ____ ____  ____ ____  ______                                     *\
**    / __// __ \/ __// __ \/ ____/    SOcos COmpiles Scala             **
**  __\_ \/ /_/ / /__/ /_/ /\_ \       (c) 2002, LAMP/EPFL              **
** /_____/\____/\___/\____/____/                                        **
**                                                                      **
** $Id$
\*                                                                      */

package scalac.util;

import java.io.*;


public class AbstractFileReader {

    /** the buffer containing the file
     */
    public byte[] buf;

    /** the current input pointer
     */
    public int bp;

    /** the file path name
     */
    public final String path;

    /** constructor
     */
    public AbstractFileReader(AbstractFile f) throws IOException {
        buf = f.read();
        bp = 0;
	path = f.getPath();
    }

    /** return byte at offset 'pos'
     */
    public byte byteAt(int pos) {
        return buf[pos];
    }

    /** read a byte
     */
    public byte nextByte() {
        return buf[bp++];
    }

    /** read some bytes
     */
    public byte[] nextBytes(int len) {
        byte[] res = new byte[len];
        System.arraycopy(buf, bp, res, 0, len);
        bp += len;
        return res;
    }

    /** read a character
     */
    public char nextChar() {
        return
            (char)(((buf[bp++] & 0xff) << 8) +
                    (buf[bp++] & 0xff));
    }

    /** read an integer
     */
    public int nextInt() {
        return  ((buf[bp++] & 0xff) << 24) +
                ((buf[bp++] & 0xff) << 16) +
                ((buf[bp++] & 0xff) <<  8) +
                (buf[bp++] & 0xff);
    }

    /** extract a character at position bp from buf
     */
    public char getChar(int bp) {
        return (char)(((buf[bp] & 0xff) << 8) + (buf[bp+1] & 0xff));
    }

    /** extract an integer at position bp from buf
     */
    public int getInt(int bp) {
        return  ((buf[bp  ] & 0xff) << 24) +
                ((buf[bp+1] & 0xff) << 16) +
                ((buf[bp+2] & 0xff) << 8) +
                 (buf[bp+3] & 0xff);
    }

    /** extract a long integer at position bp from buf
     */
    public long getLong(int bp) {
        return ((long)(getInt(bp)) << 32) + (getInt(bp + 4) & 0xffffffffL);
    }

    /** extract a float at position bp from buf
     */
    public strictfp float getFloat(int bp) {
        return Float.intBitsToFloat(getInt(bp));
    }

    /** extract a double at position bp from buf
     */
    public strictfp double getDouble(int bp) {
        return Double.longBitsToDouble(getLong(bp));
    }

   /** skip next 'n' bytes
    */
    public void skip(int n) {
        bp += n;
    }
}

File: src/compiler/scala/tools/nsc/symtab/classfile/AbstractFileReader.scala
/*
 * Scala (https://www.scala-lang.org)
 *
 * Copyright EPFL and Lightbend, Inc. dba Akka
 *
 * Licensed under Apache License 2.0
 * (http://www.apache.org/licenses/LICENSE-2.0).
 *
 * See the NOTICE file distributed with this work for
 * additional information regarding copyright ownership.
 */

package scala.tools.nsc
package symtab
package classfile

import java.io.{ByteArrayInputStream, DataInputStream}
import java.lang.Double.longBitsToDouble
import java.lang.Float.intBitsToFloat

import scala.tools.nsc.io.AbstractFile

/**
 * This class reads files byte per byte. Only used by ClassFileParser
 *
 * @author Philippe Altherr
 */
final class AbstractFileReader(val buf: Array[Byte]) extends DataReader {
  @deprecated("Use other constructor", "2.13.0")
  def this(file: AbstractFile) = this(file.toByteArray)

  /** the current input pointer
   */
  var bp: Int = 0

  /** read a byte
   */
  @throws(classOf[IndexOutOfBoundsException])
  def nextByte: Byte = {
    val b = buf(bp)
    bp += 1
    b
  }

  /** read some bytes
   */
  def nextBytes(len: Int): Array[Byte] = { // used in ide
    bp += len
    buf.slice(bp - len, bp)
  }

  /** read a character
   */
  def nextChar: Char =
    (((nextByte & 0xff) << 8) + (nextByte & 0xff)).toChar

  /** read an integer
   */
  def nextInt: Int =
    ((nextByte & 0xff) << 24) + ((nextByte & 0xff) << 16) +
    ((nextByte & 0xff) <<  8) +  (nextByte & 0xff)

  /** extract a byte at position bp from buf
   */
  def getByte(mybp: Int): Byte =
    buf(mybp)

  def getBytes(mybp: Int, bytes: Array[Byte]): Unit =
    System.arraycopy(buf, mybp, bytes, 0, bytes.length)

  /** extract a character at position bp from buf
   */
  def getChar(mybp: Int): Char =
    (((getByte(mybp) & 0xff) << 8) + (getByte(mybp+1) & 0xff)).toChar

  /** extract an integer at position bp from buf
   */
  def getInt(mybp: Int): Int =
    ((getByte(mybp) & 0xff) << 24) + ((getByte(mybp + 1) & 0xff) << 16) +
    ((getByte(mybp + 2) & 0xff) << 8) + (getByte(mybp + 3) & 0xff)

  /** extract a long integer at position bp from buf
   */
  def getLong(mybp: Int): Long =
    (getInt(mybp).toLong << 32) + (getInt(mybp + 4) & 0xffffffffL)

  /** extract a float at position bp from buf
   */
  def getFloat(mybp: Int): Float = intBitsToFloat(getInt(mybp))

  /** extract a double at position bp from buf
   */
  def getDouble(mybp: Int): Double = longBitsToDouble(getLong(mybp))

  def getUTF(mybp: Int, len: Int): String =
    new DataInputStream(new ByteArrayInputStream(buf, mybp, len)).readUTF

  /** skip next 'n' bytes
   */
  def skip(n: Int): Unit = { bp += n }
}
